File: web/plugins/multiply.js
import { WebPlugin } from "../engine/plugin.js";

export class MultiplyPlugin extends WebPlugin {
    constructor() {
        super("掛ける");
    }

    async execute(line, context) {
        const match = line.match(/^掛ける\s+(.+?)\s+(.+)$/);

        if (!match) {
            throw new Error(`掛けるの書式が正しくありません: ${line}`);
        }

        const [, target, expression] = match;

        const value = context.parser.parseNumber(
            expression,
            context.runtime
        );

        if (context.runtime.exists(target)) {
            const current = context.parser.parseNumber(
                target,
                context.runtime
            );

            context.runtime.set(target, current * value);
            return;
        }

        const current = context.parser.parseNumber(
            target,
            context.runtime
        );

        context.runtime.write(current * value);
    }
}

File: pc/plugins/multiply.py
import re
from core.plugin import Plugin


class MultiplyPlugin(Plugin):
    def translate(self, line):
        match = re.fullmatch(
            r'掛ける\s+(.+?)\s+(.+)',
            line
        )

        if not match:
            return None

        target = match.group(1).strip()
        value = match.group(2).strip()

        if re.fullmatch(r'-?\d+(?:\.\d+)?', target):
            return f"print(({target}) * ({value}))"

        if re.fullmatch(r'[A-Za-z_][A-Za-z0-9_]*|[^\W\d]\w*', target):
            return f"{target} *= {value}"

        return f"print(({target}) * ({value}))"